Expo Router "Getting Started" guide (#35463)
expo/expo: docs/pages/router/basics/navigation.mdx

---
title: Navigating bewteen pages
description: Learn the different ways to link to and navigate to pages in Expo Router.
sidebar_title: Navigation
---

import { FileTree } from '~/ui/components/FileTree';

Once you have a few pages in your app and their layouts setup, it's time to start navigating between them. Navigation in Expo Router works a lot like React Navigation, but with all pages having a URL by default, we can create links and use these URL's to move about our app using familiar web patterns.

## Native navigation basics with `useRouter`

Like in React Navigation, you can call a function from an `onPress` handler to navigate to another page. In Expo Router, you can use the `useRouter` hook to access navigation functions:

```tsx
import { useRouter } from 'expo-router';

export default function Home() {
  const router = useRouter();

  return <Button onPress={() => router.navigate('/about')}>Go to About</Button>;
}
```

> Expo Router apps default to stack navigation, where navigating to a new route pushes a screen onto a stack, and backing out of that route pops it off the stack. Usually, you would want to use the `router.navigate` function. This will either push a new page onto the stack or unwinding to an existing route on the stack. However, you can also call `router.push` to explicitly push a new page onto the stack, `router.back` to go back to the previous page, or `router.replace` to replace the current page on the stack.

Reviewer: ```suggestion
Expo Router apps default to stack navigation, where navigating to a new route pushes a screen onto a stack, and backing out of that route pops it off the stack. Usually, you would want to use the `router.navigate` function. This will either push a new page onto the stack or unwind to an existing route on the stack. However, you can also call `router.push` to explicitly push a new page onto the stack, `router.back` to go back to the previous page, or `router.replace` to replace the current page on the stack.
```

> With Expo Router, you refer to pages by their URL, or their position relative to the **app** folder. Check out the following file structure and how you would navigate to each page:

Reviewer: ```suggestion
With Expo Router, you refer to pages by their URL, or their position relative to the **app** directory. Check out the following file structure and how you would navigate to each page:
```

<FileTree
  files={[

>     ['app/index.tsx', '`router.navigate("/")`'],

Reviewer: Yes, they are already in code syntax format. We should just remove backticks from here.

    ['app/about.tsx', '`router.navigate("/about")`'],
    ['app/profile/index.tsx', '`router.navigate("/profile")`'],
    ['app/profile/friends.tsx', '`router.navigate("/profile/friends")`'],
  ]}
/>

> ## Links and Buttons

Reviewer: ```suggestion
## Links and buttons
```

> However, the typical way to link to a page in Expo Router is to use links, just like web apps. Expo Router has `Link` components for navigating between pages, where the `href` is the same route you would use in `router.navigate`:

Reviewer: ```suggestion
The typical way to link to a page in Expo Router is to use links like web apps. Expo Router has a `Link` component for navigating between pages, where the `href` is the same route you would use in `router.navigate`:
```

There's a singular `Link` keyword.

{/* prettier-ignore */}
```tsx app/index.tsx
import { View } from 'react-native';

> /* @info Import the <strong>Link</strong> React component from <strong>expo-router</strong> */

Reviewer: ```suggestion
/* @info Import the <CODE>Link</CODE> React component from <CODE>expo-router</CODE>. */
```

Instead of `strong`, let's use `CODE` for annotating they keywords here correctly.

import { Link } from 'expo-router';
/* @end */

export default function Page() {
  return (
    <View>

>       /* @info Tapping this will link to the <strong>about</strong> page */
>       <Link href="/about">About</Link>
>     </View>

Reviewer: ```suggestion
      /* @info Tapping this will link to the <strong>about</strong> page. */
      <Link href="/about">About</Link>
      /* @end */
    </View>
```

Missing closing tag for the annotation here.

  );
}
```

By default, Links can only wrap `Text` components. You can use `Pressable` or other components that support `onPress` and `onClick` props inside a link with the `asChild` prop:

{/* prettier-ignore */}
```tsx
import { Pressable, Text } from 'react-native';
import { Link } from 'expo-router';

export default function Page() {
  return (

>     /* @info The <strong>onPress</strong> event that navigates to <strong>/other</strong> will be passed to <strong>Pressable</strong> */

Reviewer: ```suggestion
    /* @info The <CODE>onPress</CODE> event that navigates to <CODE>/other</CODE> will be passed to <CODE>Pressable</CODE>. */
```

    <Link href="/other" asChild>
    /* @end */
      <Pressable>
        <Text>Home</Text>
      </Pressable>
    </Link>
  );
}
```

## Relative routes

> You don't always have to use the absolute path to a route. Using paths that start with `./` (for the current folder) or `../` (for the parent folder) will navigate relative to the current route.

Reviewer 2: I didn't know you could do that! I notice that `<Link>` also has a `relativeToDirectory` prop. I'm curious what the best practice is for navigating using a relative path? @marklawlor

> You don't always have to use the absolute path to a route. Using paths that start with `./` (for the current folder) or `../` (for the parent folder) will navigate relative to the current route.

Reviewer: ```suggestion
You don't always have to use the absolute path to a route. Using paths that start with `./` (for the current directory) or `../` (for the parent directory) will navigate relative to the current route.
```

A relative URL is a URL prefix with `./`, such as `./article`, or `./article/`. Relative URLs are resolved relative to the current rendered screen.

```tsx
<Link href="./article">Go to article</Link>
```

```ts
router.navigate('./article');
```

## Dynamic routes and URL parameters

Dynamic routes can be linked to with their full URL, or by passing a `params` object.

> Consider this file structure:

Reviewer: ```suggestion
Consider the following file structure:
```

<FileTree files={[['app/user/[id].tsx']]} />

Each of these links will navigate to the same page:

{/* prettier-ignore */}
```tsx app/index.tsx

> /* @info Import the <strong>Link</strong> React component from <strong>expo-router</strong> */

Reviewer: ```suggestion
/* @info Import the <CODE>Link</CODE> React component from <CODE>expo-router</CODE>. */
```

import { Link } from 'expo-router';
/* @end */
import { View, Pressable } from 'react-native';

export default function Page() {
  return (
    <View>
      <Link
        href="/user/bacon">
        View user (id inline)
      </Link>
      <Link
        href={{

>           /* @info Navigate to <strong>/user/bacon</strong> */

Reviewer 2: Is this highlight intentional? Should it highlight all 3 options?

>           /* @info Navigate to <strong>/user/bacon</strong> */

Reviewer: ```suggestion
          /* @info Navigate to <CODE>/user/bacon</CODE>. */
```

          pathname: '/user/[id]',
          params: { id: 'bacon' }
          /* @end */
        }}>
        View user (id in params in href)
      </Link>
      <Pressable onPress={() => router.navigate('/user', { id: 'bacon' })}>
        <Text>View user (imperative)</Text>
      </Pressable>
    </View>
  );
}
```

### Passing query parameters

You can specify query parameters in the link URL itself, or as additional parameters in the `params` object. Any parameters that don't match the name of the dynamic route variable are equivalent to query parameters.

```tsx
<Link href="/users?limit=20">View users</Link>

<Link
  href={{
    pathname: '/users',
    params: { limit: 20 }
  }}>
  View users
</Link>
```

### Using dynamic route variables and query parameters in the destination page

All variables in the link URL are accessible to the receiving page via the `useLocalSearchParams` hook. This hook returns an object with all the URL parameters, including those passed as `params`.

> So, if you have a link like this:

Reviewer: ```suggestion
For example, if you have a link like this:
```

```

Then you can read the parameters on the other end like this:

```tsx
import { useLocalSearchParams } from 'expo-router';
import { View, Text } from 'react-native';

export default function Users() {
  const { id, limit } = useLocalSearchParams();

  return (
    <View>
      <Text>User ID: {id}</Text>
      <Text>Limit: {limit}</Text>
    </View>
  );
}
```

### Updating query parameters without navigating

Query parameters can be updated without navigating to a new page. This can be done with a `Link` that uses the same URL as the current page, but with updated query parameters, or imperatively.

```tsx
<Link href="/users?limit=50">View more users</Link>

<Pressable onPress={() => router.setParams({ limit: 50 })}>
  <Text>View more users</Text>
</Pressable>
```

## Redirects

You can immediately redirect to another route from a page or layout with the `Redirect` component. This functions like the `replace` imperative navigation function. A redirect will navigate to the new route without rendering the current page.

```tsx
import { Redirect } from 'expo-router';

export default function Page() {
  return <Redirect href="/about" />;
}
```

## Deep links

> Deep linking is when a URL opens a specific page in your app. Expo Router supports deep linking by default, so you can link to any page in your app with a URL from outside of your app, just like you would inside your app with `Link`. This is especially useful for sharing links to specific pages in your app.

Reviewer: ```suggestion
Deep linking is when a URL opens a specific page in your app. Expo Router supports deep linking by default, so you can link to any page in your app with a URL from outside of your app, as you would inside your app with `Link`. This is especially useful for sharing links to specific pages in your app.
```

> On web, deep linking is as simple as navigating to that specific URL in your web browser. On mobile, you define a `scheme` in your [app config](), and this becomes the prefix for deep links into your app.

Reviewer: ```suggestion
On web, deep linking is as simple as navigating to that specific URL in your web browser. On mobile, you define a `scheme` in your [app config](/workflow/configuration/) file, and this becomes the prefix for deep links into your app.
```

> Assuming your `scheme` is `myapp`, here's some examples of how you would link to a page in your app from a web page or another app:

Reviewer: ```suggestion
Assuming your `scheme` is `myapp`, here are some examples of how you would link to a page in your app from a web page or another app:
```

<FileTree
  files={[
    ['app/about.tsx', 'myapp://about'],
    ['app/profile/index.tsx', 'myapp://profile'],
    ['app/users/[username].tsx', 'myapp://users/evanbacon'],
  ]}
/>

> With app links and universal links, you can also link into your app with an https URL. [Learn more]()

Reviewer: ```suggestion
With app links and universal links, you can also link to your app with an `https` URL. For more information, see [Universal linking](/linking/overview/#universal-linking).
```

## Initial routes

> When opening a deep link to a page in your app, you will likely want back navigation to work just as if the user navigated to the page from your home page. To do this, you can specify an `initalRouteName` configuration, which defines the page in a layout that should be loaded prior to the deep linked page.

Reviewer: ```suggestion
When opening a deep link to a page in your app, you will likely want back navigation to work as if the user navigated to the page from your home page. To do this, you can specify an `initalRouteName` configuration, which defines the page in a layout that should be loaded prior to the deep linked page.
```

Consider the following file structure:

<FileTree
  files={[
    ['app/index.tsx'],
    ['app/stack/index.tsx'],
    ['app/stack/second.tsx'],

>     ['app/stack/_layout_.tsx'],

Reviewer: ```suggestion
    ['app/stack/_layout.tsx'],
```

  ]}
/>

`stack` is a stack navigator, and `/stack/index` is always the first route in the stack.

> To ensure that `/stack/index` is always loaded first, even if the user deep links to `/stack/second`, you can set the `initialRouteName` in **app/stack/_layout_.tsx**:

Reviewer: ```suggestion
To ensure that `/stack/index` is always loaded first, even if the user deep links to `/stack/second`, you can set the `initialRouteName` in **app/stack/_layout.tsx**:
```

```tsx
export const unstable_settings = {
  // Ensure any route can link back to `/`
  initialRouteName: 'index',
};
```

> By default, the `initialRouteName` is only considered when deep linking, and not during navigation within your app. However, you can use the `withAnchor` prop on `Link` to force the initial route to be loaded when navigating directly into another stack inside your app.

Reviewer: ```suggestion
By default, the `initialRouteName` is only considered when deep linking and not during navigation within your app. However, you can use the `withAnchor` prop on `Link` to force the initial route to be loaded when navigating directly into another stack inside your app.
```

---

> title: Navigating bewteen pages

Reviewer 3: ```suggestion
title: Navigating between pages
```

description: Learn the different ways to link to and navigate to pages in Expo Router.
sidebar_title: Navigation
---

import { FileTree } from '~/ui/components/FileTree';

> Once you have a few pages in your app and their layouts setup, it's time to start navigating between them. Navigation in Expo Router works a lot like React Navigation, but with all pages having a URL by default, we can create links and use these URL's to move about our app using familiar web patterns.

Reviewer 4: ```suggestion
Once you have a few pages in your app and their layouts setup, it's time to start navigating between them. Navigation in Expo Router works a lot like React Navigation, but with all pages having a URL by default, we can create links and use these URLs to move about our app using familiar web patterns.
```

## Native navigation basics with `useRouter`

Like in React Navigation, you can call a function from an `onPress` handler to navigate to another page. In Expo Router, you can use the `useRouter` hook to access navigation functions:

```tsx
import { useRouter } from 'expo-router';

export default function Home() {
  const router = useRouter();

  return <Button onPress={() => router.navigate('/about')}>Go to About</Button>;
}
```

> Expo Router apps default to stack navigation, where navigating to a new route pushes a screen onto a stack, and backing out of that route pops it off the stack. Usually, you would want to use the `router.navigate` function. This will either push a new page onto the stack or unwind to an existing route on the stack. However, you can also call `router.push` to explicitly push a new page onto the stack, `router.back` to go back to the previous page, or `router.replace` to replace the current page on the stack.

Reviewer 4: Should router.dismiss() get a mention too?

Author: Now that we have the API reference, we were intending to pare back being exhaustive in the guides themselves. However, I added a callout to the API reference for this and. the `Link` section to make these other functions easier to find from here.

```tsx
import { Redirect } from 'expo-router';

export default function Page() {
  return <Redirect href="/about" />;
}
```

## Deep links

Deep linking is when a URL opens a specific page in your app. Expo Router supports deep linking by default, so you can link to any page in your app with a URL from outside of your app, as you would inside your app with `Link`. This is especially useful for sharing links to specific pages in your app.

On web, deep linking is as simple as navigating to that specific URL in your web browser. On mobile, you define a `scheme` in your [app config](/workflow/configuration/) file, and this becomes the prefix for deep links into your app.

Assuming your `scheme` is `myapp`, here are some examples of how you would link to a page in your app from a web page or another app:

<FileTree
  files={[
    ['app/about.tsx', 'myapp://about'],
    ['app/profile/index.tsx', 'myapp://profile'],
    ['app/users/[username].tsx', 'myapp://users/evanbacon'],
  ]}
/>

With app links and universal links, you can also link to your app with an `https` URL. For more information, see [Universal linking](/linking/overview/#universal-linking).

## Initial routes

> When opening a deep link to a page in your app, you will likely want back navigation to work as if the user navigated to the page from your home page. To do this, you can specify an `initalRouteName` configuration, which defines the page in a layout that should be loaded prior to the deep linked page.

Reviewer 3: ```suggestion
When opening a deep link to a page in your app, you will likely want back navigation to work as if the user navigated to the page from your home page. To do this, you can specify an `initialRouteName` configuration, which defines the page in a layout that should be loaded prior to the deep linked page.
```

Typo initalRouteName -> initialRouteName
